configure.ac

```diff
@@ -44,59 +58,122 @@ AC_PROG_CPP
+GX_FC_QUAD_PRECISION()
 
-# Require netCDF.
-#AC_CHECK_FUNC([nf_open], [], [AC_MSG_ERROR([NetCDF Fortran library required to build FMS])])
+# Check if Fortran compiler supports reading namelists from internal files
+GX_FC_INTERNAL_FILE_NML()
 
-# Do we want to try and set Fortran flags for the user?
-if test $enable_fortran_flag_setting = "yes"; then
-   # Make sure the compiler is seeing this as free-formatted, not
-   # fixed-formatted, fortran code.
-   AC_FC_FREEFORM()
+if test $with_openmp = yes && test $enable_setting_flags = yes; then
+  # Check if the compiler needs special OpenMP flags
+  AC_OPENMP()
+fi
+AC_LANG_POP(Fortran)
 
-   # Check that long lines of Fortran code can be handled. This will add
-   # necessary fortran flags.
-   AC_FC_LINE_LENGTH([unlimited])
+# We passed all the tests.  Set the required defines.
+AC_DEFINE([use_netCDF], [1], [This is required for the library to build])
+if test $with_mpi = yes; then
+  AC_DEFINE([use_libMPI], [1], [This is required for the library to build])
 fi
 
-# These defines are required for the build.
-AC_DEFINE([use_netCDF], [1])
-AC_DEFINE([use_libMPI], [1])
+# Set any required compile flags.  This will not be done if the user wants to
+# define all their own flags.
+if test $enable_setting_flags = yes; then
+  # Make sure the compiler is seeing this as free-formatted, not
+  # fixed-formatted, fortran code.
+  AC_FC_FREEFORM()
+
+  # Check that long lines of Fortran code can be handled. This will add
+  # necessary fortran flags.
+  AC_FC_LINE_LENGTH([unlimited])
+
+  # Will we build with default 64-bit reals in Fortran, or do mixed mode?
+  if test $enable_mixed_mode = yes; then
+    GX_FC_DEFAULT_REAL_KIND4_FLAG([dnl
+      FCFLAGS="$FCFLAGS $FC_DEFAULT_REAL_KIND8_FLAG"])
+    AC_DEFINE([OVERLOAD_R4], [1], [Set to overload the R4 Fortran routines])
+    AC_DEFINE([OVERLOAD_R8], [1], [Set to overload the R8 Fortran routines])
+  else
+    GX_FC_DEFAULT_REAL_KIND8_FLAG([dnl
+      FCFLAGS="$FCFLAGS $FC_DEFAULT_REAL_KIND8_FLAG"])
+  fi
+
+  # Add Cray Pointer support flag
+  if test ! -z "$FC_CRAY_POINTER_FLAG"; then
+    FCFLAGS="$FCFLAGS $FC_CRAY_POINTER_FLAG"
+  fi
+
+  # Add OpenMP flags
+  if test ! -z "$OPENMP_CFLAGS"; then
+    CFLAGS="$CFLAGS $OPENMP_CFLAGS"
+    CPPFLAGS="$CPPFLAGS $OPENMP_CFLAGS"
+  fi
+  if test ! -z "$OPENMP_FCFLAGS"; then
+    FCFLAGS="$FCFLAGS $OPENMP_FCFLAGS"
+  fi
+fi
 
+# Settings based on the system and hardware
 # Define __APPLE__ macro if on a Apple/Mac OS X
 AS_IF([test "x$target_vendor" = "xapple"], [AC_DEFINE([__APPLE__], [ ])])
-
-# Define an AM_CONDITIONAL to determine if you are on a CRAY
-AM_CONDITIONAL([CRAY], [test `env | grep CRAY | wc -l` -gt 0])
-
 # Define an AM_CONDITIONAL to determine if you are on an Apple/Mac OS X
 AM_CONDITIONAL([APPLE], [test "x$target_vendor" = "xapple"])
+# Define an AM_CONDITIONAL to determine if you are on a CRAY
+AM_CONDITIONAL([CRAY], [test `env | grep CRAY | wc -l` -gt 0])
```

Comment:
Why do you need to know if you are on a cray?

Reply:
The static libraries are still the default on the Cray machines at NCRC.  The macro is needed by the various unit test Makefiles to create the correct library link specification.  The macro name could be changed to be more explicit, such as LINK_STATIC_LIB.  But to date, this is a machine-specific test.

Reply:
OK, this is incorrect for several reasons and should not be merged.

In this case, tell the builders on the Cray machines at NCRC to use --disable-shared. (Also tell them to fix their installs; it's not hard.)

So no need to change the build system, this is already fully supported. You're done! ;-)

Secondly, even if this was necessary, there is no need to make this machine specific. Instead, always define a more explicit macro. Otherwise you are encouraging lazy programming and brittle build systems. (Because other programmers will see that you have CRAY defined, and will start using it. Now your changes, and theirs, will not work for non-cray systems that share the build characteristic you are dealing with. Everything will become machine-specific and non-portable.)

After defining a specific macro to cover the build feature you need to determine, go one step further, and write an autoconf test which works on all machines and resolve the question of how to set your explicit macro. (I can help here but Seth is also well-positioned to tell how to do this.)

Then you will have a fully-portable solution that will work on new environments, not just one cray at NCRC. (If these changes are merged, it will break shared library installs for all other crays, which have installed shared libraries correctly. That's certainly not a good idea.)

Please never check in any part of this build system that relies on machine-specific rules. You will be breaking portability for NOAA and other users, including at NCRC, when they get a new Cray or fix installs on their existing one.